### PR TITLE
iox-#1394 Prevent destruction of NewType derived classes by base class pointer

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

The `NewType` makes it easy to circumvent the primitive obsession and prevents your mars lander crashing into a planet by deriving from `NewType` and specifying policies which traits of a primitive type should be used.
This leads to the possibility to destruct the derived classes via an pointer to the `NewType`. Since the `NewType` cannot have a virtual destructor (due to usage in shared memory) this must be prevented by making the destructor of `NewType` and the policy classes protected.

See also https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-virtual

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1394

### CODE REVIEW

**Review by elfenpiff — APPROVED** [review]

**Review by elfenpiff — CHANGES_REQUESTED** [review]
Oh I realized the change itself is correct and I agree with it but in one month we will have forgotten why we added this.
Could you please add a detailed comment why this is required.

In the end you can copy & paste the note for the reviewer from the PR.

Oh and you were faster in #1415 were I left the identical comment. Could you please add the comment in this PR.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/assignment.hpp [hunk 1]
```diff
@@ -26,21 +26,29 @@ namespace newtype
 template <typename T>
 struct CopyAssignable
 {
+  protected:
+    ~CopyAssignable() = default;
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/assignment.hpp:30]: On a second thought, this is also something that clang-tidy will check and there is always the git history and therefore I would omit the comment

**Review by elfenpiff — DISMISSED** [review]

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 2]
```diff
@@ -139,3 +139,37 @@
...
+   // compile time error when Foo and Bar are mixed up
+   struct Foo : public iox::cxx::NewType<uint64_t, iox::cxx::newtype::ConstructByValueCopy>
+   {
+       using ThisType::ThisType;
+   };
+
+   struct Bar : public iox::cxx::NewType<uint64_t, iox::cxx::newtype::ConstructByValueCopy>
+   {
+       using ThisType::ThisType;
+   };
```
Comment (elBoberido) [inline doc/website/release-notes/iceoryx-unreleased.md:174]: What do you think about such a macro?
```cpp
#define IOX_NEW_TYPE(TypeName, Type, ...)                            \
    struct TypeName : public iox::cxx::NewType<Type, __VA_ARGS__>    \
    {                                                                \
        using ThisType::ThisType;                                    \
        using ThisType::operator=;                                   \
                                                                     \
        TypeName() noexcept = default;                               \
        TypeName(const TypeName&) noexcept = default;                \
        TypeName(TypeName&&) noexcept = default;                     \
        TypeName& operator=(const TypeName&) noexcept = default;     \
        TypeName& operator=(TypeName&&) noexcept = default;          \
    };

IOX_NEW_TYPE(Foo, uint64_t, iox::cxx::newtype::ConstructByValueCopy)
IOX_NEW_TYPE(Bar, uint64_t, iox::cxx::newtype::ConstructByValueCopy)
```

If you like it, I can create an issue an implement it.

**Review by mossmaurice — APPROVED** [review]
LGTM, thanks!

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/comparable.hpp [hunk 3]
```diff
@@ -36,6 +36,9 @@ struct Comparable
     {
         return !(lhs == rhs);
     }
+
+  protected:
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/comparable.hpp:40]: Copyright header :upside_down_face:

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/assignment.hpp [hunk 4]
```diff
@@ -26,21 +26,29 @@ namespace newtype
 template <typename T>
 struct CopyAssignable
 {
+  protected:
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/assignment.hpp:29]: Copyright header :upside_down_face:

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/newtype.hpp [hunk 5]
```diff
@@ -75,6 +75,11 @@ class NewType : public Policies<NewType<T, Policies...>>...
   protected:
     NewType(newtype::internal::ProtectedConstructor_t, const T& rhs) noexcept;
 
+    /// @note Since `using Foo = NewType<int>` and `using Bar = NewType<int>` result
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/newtype.hpp:78]: Copyright header :upside_down_face:

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/sortable.hpp [hunk 6]
```diff
@@ -46,6 +46,9 @@ struct Sortable
     {
         return internal::newTypeAccessor(lhs) >= internal::newTypeAccessor(rhs);
     }
+
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/sortable.hpp:49]: Copyright header :upside_down_face:

File: doc/website/release-notes/iceoryx-unreleased.md (hunk 2 above)
Comment (elfenpiff) [reply doc/website/release-notes/iceoryx-unreleased.md:174]: I think we do not have a use case in iceoryx where such a macro would be used at the moment. I think it is quite often the case that we inherit from newtype and add some additional methods, see `UniquePortId` or customize the constructor see `UniqueId`.

Comment (elBoberido) [reply doc/website/release-notes/iceoryx-unreleased.md:184]: @elfenpiff maybe it is not used because the threshold is higher than necessary ;)

Comment (elfenpiff) [reply doc/website/release-notes/iceoryx-unreleased.md:184]: Then add this macro :D

**Review by elfenpiff — APPROVED** [review]

**Review by mossmaurice — APPROVED** [review]

**Pull request merged**